### PR TITLE
SITL: added stack checking on threads 

### PR DESCRIPTION
**Pull request by tridge**
Useful for checking stack sizes

### CODE REVIEW

File: libraries/AP_HAL_SITL/Scheduler.cpp [hunk 1]
```diff
@@ -209,29 +216,98 @@ void Scheduler::stop_clock(uint64_t time_usec)
...
+  check for stack overflow
+ */
+void Scheduler::check_thread_stacks(void)
+{
+    WITH_SEMAPHORE(_thread_sem);
+    for (struct thread_attr *p=threads; p; p=p->next) {
+        const uint8_t ncheck = 8;
+        for (uint8_t i=0; i<ncheck; i++) {
+            if (p->stack_min[i] != stackfill) {
+                AP_HAL::panic("stack overflow in thread %s\n", p->name);
```
Comment (peterbarker) [inline libraries/AP_HAL_SITL/Scheduler.cpp]: There's general agreement amongst the HALs that the `\n` is inferred - ChibiOS is an exception.

File: libraries/AP_HAL_SITL/Scheduler.cpp [hunk 2]
```diff
@@ -209,29 +216,98 @@ void Scheduler::stop_clock(uint64_t time_usec)
...
+    }
+    if (pthread_create(&thread, &a->attr, thread_create_trampoline, a) != 0) {
+        goto failed;
+    }
+    a->next = threads;
+    threads = a;
     return true;
+
+failed:
+    if (a->stack) {
```
Comment (peterbarker) [inline libraries/AP_HAL_SITL/Scheduler.cpp]: free accepts nullptr

File: libraries/AP_HAL_SITL/Scheduler.cpp [hunk 3]
```diff
@@ -209,29 +216,98 @@ void Scheduler::stop_clock(uint64_t time_usec)
...
+    struct thread_attr *a = new struct thread_attr;
+    if (!a) {
         return false;
     }
+    // take a copy of the MemberProc, it is freed after thread exits
+    a->f = (AP_HAL::MemberProc *)malloc(sizeof(proc));
+    if (!a->f) {
+        goto failed;
+    }
+    a->stack = memalign(4096, alloc_stack);
```
Comment (peterbarker) [inline libraries/AP_HAL_SITL/Scheduler.cpp]: Rumour has it `memalign` is deprecated in favour of `posix_memalign`

**Comment by lvale** [comment]
@tridge 
SITL on MacOS broken with this
```

Sparkle:ArduPilot Luis$ ./Tools/autotest/sim_vehicle.py -v ArduCopter -f copter -L TagusPark
SIM_VEHICLE: Start
WARNING: no config for frame (copter)
SIM_VEHICLE: Killing tasks
SIM_VEHICLE: Starting up at 38.740195,-9.301843,151,0 (TagusPark)
SIM_VEHICLE: WAF build
SIM_VEHICLE: Configure waf
SIM_VEHICLE: "/Users/Luis/src/ardupilot/modules/waf/waf-light" "configure" "--board" "sitl"
Setting top to                           : /Users/Luis/src/ardupilot 
Setting out to                           : /Users/Luis/src/ardupilot/build 
Autoconfiguration                        : enabled 
Setting board to                         : sitl 
Checking for 'g++' (C++ compiler)        : not found 
Checking for 'clang++' (C++ compiler)    : /usr/bin/clang++ 
Checking for 'gcc' (C compiler)          : not found 
Checking for 'clang' (C compiler)        : /usr/bin/clang 
Checking for c flags '-MMD'              : yes 
Checking for cxx flags '-MMD'            : yes 
Disabling SLP for clang++
Checking for HAVE_CMATH_ISFINITE         : yes 
Checking for HAVE_CMATH_ISINF            : yes 
Checking for HAVE_CMATH_ISNAN            : yes 
Checking for NEED_CMATH_ISFINITE_STD_NAMESPACE : yes 
Checking for NEED_CMATH_ISINF_STD_NAMESPACE    : yes 
Checking for NEED_CMATH_ISNAN_STD_NAMESPACE    : yes 
Checking for header endian.h                   : not found 
Checking for header byteswap.h                 : not found 
Checking for program 'python'                  : /usr/local/opt/python@2/bin/python2.7 
Checking for python version >= 2.7.0           : 2.7.15 
Checking for program 'python'                  : /usr/local/opt/python@2/bin/python2.7 
Checking for python version >= 2.7.0           : 2.7.15 
Source is git repository                       : yes 
Update submodules                              : yes 
Checking for program 'git'                     : /usr/local/bin/git 
Checking for program 'size'                    : /usr/bin/size 
Benchmarks                                     : disabled 
Unit tests                                     : enabled 
Checking for program 'rsync'                   : /usr/bin/rsync 
'configure' finished successfully (1.518s)
SIM_VEHICLE: Building
SIM_VEHICLE: "/Users/Luis/src/ardupilot/modules/waf/waf-light" "build" "--target" "bin/arducopter"
Waf: Entering directory `/Users/Luis/src/ardupilot/build/sitl'
[383/506] Compiling libraries/AP_HAL_SITL/Scheduler.cpp
[438/506] Compiling ArduCopter/ekf_check.cpp
[439/506] Compiling ArduCopter/mode_flip.cpp
[440/506] Compiling ArduCopter/inertia.cpp
[441/506] Compiling ArduCopter/UserCode.cpp
[442/506] Compiling ArduCopter/mode_throw.cpp
[443/506] Compiling ArduCopter/mode_smart_rtl.cpp
[444/506] Compiling ArduCopter/Copter.cpp
../../libraries/AP_HAL_SITL/Scheduler.cpp:266:16: fatal error: use of undeclared identifier 'memalign'
    a->stack = memalign(4096, alloc_stack);
               ^
1 error generated.

Waf: Leaving directory `/Users/Luis/src/ardupilot/build/sitl'
Build failed
 -> task in 'objs/AP_HAL_SITL' failed (exit status 1): 
	{task 4580087640: cxx Scheduler.cpp -> Scheduler.cpp.0.o}
 (run with -v to display more information)
SIM_VEHICLE: Build failed
SIM_VEHICLE: Killing tasks
Sparkle:ArduPilot Luis$ 
Sparkle:ArduPilot Luis$ 

```

**Pull request merged**